hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/DataInputNavigableJsonAdapter.java

```diff
@@ -88,34 +90,78 @@ private boolean isQuote() throws IOException {
         return input.readByte() == '"';
     }
 
-    private static class UTF8Reader extends Reader {
+    static class UTF8Reader extends Reader {
 
-        private final DataInput input;
+        private final CharsetDecoder utf8Decoder;
+        private final ByteBuffer inputBuffer;
+        // required to support read() for multibyte characters
+        private boolean hasLeftoverChar;
+        private int leftoverChar;
 
-        UTF8Reader(DataInput input) {
-            this.input = input;
+        UTF8Reader(BufferObjectDataInput input) {
+            byte[] data = obtainBytes(input);
+            inputBuffer = ByteBuffer.wrap(data);
+            inputBuffer.position(input.position());
+            this.utf8Decoder = Bits.UTF_8.newDecoder();
+        }
+
+        // default read() implementation does not handle multibyte chars well
+        @Override
+        public int read() throws IOException {
+            if (hasLeftoverChar) {
+                hasLeftoverChar = false;
+                return leftoverChar;
```

Comment:
If `read(char[]...)` is called after first calling `read()`, the left-over character from `read()` may get lost (or returned later, out of order, if `read()` is called again).